tests/test_energy_overflows.py

```diff
@@ -0,0 +1,282 @@
+        unbound_pot = gpu_pot.unbound_impl
+        bound_pot = gpu_pot.bind(params).bound_impl
+        _, _, selective_energy = unbound_pot.execute_selective(x0, params, box, False, False, True)
+        _, bound_energy = bound_pot.execute(x0, box)
+        verify_energies(selective_energy, bound_energy)
+
+        # Check that all of the energies out of a batch are identical
+        _, _, energies = unbound_pot.execute_selective_batch([x0] * 2, [params] * 2, [box] * 2, False, False, True)
+        for energy in energies.reshape(-1):
+            verify_energies(selective_energy, energy)
+
+        fixed_energy, overflows = bound_pot.execute_fixed(x0, box)
+
+        # If there are no overflows, the fixed energy value will match
+        if overflows == 0:
+            assert FIXED_TO_FLOAT(fixed_energy) == selective_energy
+
+        return bound_energy
+
+    reference_energy = potential(x0, params, box)
+    test_energy = compute_potential_energy(potential)
+    np.testing.assert_allclose(reference_energy, test_energy, rtol=rtol, atol=atol)
+
+    # If we call the two potentials separately, we will now see that both return inf energies
+    all_pairs_energy = compute_potential_energy(all_pairs)
+    assert not np.isfinite(all_pairs_energy)
+
+    pair_list_energy = compute_potential_energy(pair_list)
+    assert not np.isfinite(pair_list_energy)
+
+    # Make sure that the overflows are actually cancelling out
+    pair_list_bound = pair_list.to_gpu(precision).bind(params).bound_impl
+    all_pairs_bound = all_pairs.to_gpu(precision).bind(params).bound_impl
+
+    fixed_all_pairs_energy, all_pairs_overflows = all_pairs_bound.execute_fixed(x0, box)
+    fixed_pair_list_energy, pair_list_overflows = pair_list_bound.execute_fixed(x0, box)
+    assert all_pairs_overflows != 0
+    assert pair_list_overflows != 0
+    assert all_pairs_overflows + pair_list_overflows == 0
+
+    # The sum of the fixed point energies should be bitwise identical to the test energy
+    assert FIXED_TO_FLOAT(fixed_all_pairs_energy + fixed_pair_list_energy) == test_energy
+
+
+@pytest.mark.xfail(strict=True)
+@pytest.mark.parametrize("precision", [np.float32, np.float64])
+def test_energy_overflows_with_summation_of_energies(precision):
+
+    num_atoms = 1000
+
+    # all particles evenly spaced along x-axis, in a very large box
+    spacing = 0.75
+    x = spacing * np.array([np.arange(num_atoms), np.zeros(num_atoms), np.zeros(num_atoms)]).T
+    box = np.eye(3) * 10000.0
+    assert (np.diag(box) > spacing * num_atoms).all()
+
+    nb_params = np.ones((num_atoms, 4))
+    nb_params[:, -1] = 0.0
+
+    nonbonded = Nonbonded(
+        num_atoms=num_atoms,
+        exclusion_idxs=np.array([(0, num_atoms - 1)], dtype=np.int32),
+        scale_factors=np.zeros((1, 2)),
+        beta=2.0,
+        cutoff=1.2,
+    )
+    nonbonded_gpu = nonbonded.to_gpu(precision)
+
+    # TBD: Remove the try/finally to clean up the GPU potential, else calling reset_device
```

Comment:
Is this TBD for this PR, or future? (not sure I understand why explicit cleanup is needed in this case)

Reply:
For the future. If you don't have this, the `device_reset` will clear the memory, which will result in a failure of compute sanitizer due to it thinking the memory is lost.

I chock this up to python garbage collection not playing well with C++ objects. There is a call to the gc in the tests, but doesn't clean up this object. Doing it explicitly fixes it, and when this test starts passing we can remove the try/fnally